### PR TITLE
Rework of the mdmp plugin

### PR DESCRIPTION
**Pull request by AlexKornitzer**
The previous implementation of the MDMP plugin was missing certain functionality required for useful analysis within radare2. So I spent the weekend reworking/reimplementing the plugin.

Major changes:

- Refactored the code to remove unnecessary type definitions and converted to snake_case to match the rest of the project.
- Updated the Sections section to map the physical address to the correct location in the MDMP file if present (i.e. full dumps). If not present the paddr is set to 0.

```
idx=00 vaddr=0xff7b0000 paddr=0x018825a8 sz=929792 vsz=929792 perm=----- name=C:_Windows_System32_calc.exe
idx=01 vaddr=0x77720000 paddr=0x016d25a8 sz=1744896 vsz=1744896 perm=----- name=C:_Windows_System32_ntdll.dll
idx=02 vaddr=0x77500000 paddr=0x014b95a8 sz=1175552 vsz=1175552 perm=----- name=C:_Windows_System32_kernel32.dll
[snip]
idx=25 vaddr=0x7fefd2b0000 paddr=0x01fd95a8 sz=61440 vsz=61440 perm=----- name=C:_Windows_System32_CRYPTBASE.dll
idx=26 vaddr=0x7fefddb0000 paddr=0x022b25a8 sz=626688 vsz=626688 perm=----- name=C:_Windows_System32_clbcatq.dll
idx=27 vaddr=0x7fef62b0000 paddr=0x019655a8 sz=344064 vsz=344064 perm=----- name=C:_Windows_System32_oleacc.dll
```

- Made use of the Memory section to store all memory mappings found within the MDMP file. Again used the paddr to display the location of the memory within the file.

```
0x00110000 +0x1000 ---- paddr=0x000b55a8 RAM
0x00120000 +0xa000 ---- paddr=0x000b65a8 RAM
0x00130000 +0x2000 ---- paddr=0x000c05a8 RAM
0x00140000 +0xa000 ---- paddr=0x000c25a8 RAM
0x00150000 +0xa000 ---- paddr=0x000cc5a8 RAM
0x00160000 +0x10000 ---- paddr=0x000d65a8 RAM
...
```

Issues:
- Searching or analysing the memory within the MDMP file is not possible without setting `io.va=false`.
- The `r_str_utf16_to_utf8()` function seems to remove `\` characters?

TODO:
- Full Parsing - Currently parsing is only performed to depth one if the data is not going to be presented in radare2.
- Find a way to present the debugging information to the radare2 debugger.
- Add non-standard streams.
- Test on more dumps than just basic/full of calc.exe

Final Notes: Hopefully I have not done anything too catastrophic to this plugin, and that people will find the changes useful. If any bugs are encountered from the implementation send them my way.

### CODE REVIEW

File: libr/bin/format/mdmp/mdmp.c [hunk 1]
```diff
@@ -1,371 +1,296 @@
...
-}
+static bool r_bin_mdmp_init_hdr(struct r_bin_mdmp_obj *obj) {
+  obj->hdr = (struct minidump_header *)obj->b->buf;
 
-int r_bin_mdmp_init(struct r_bin_mdmp_obj *obj) {
-	if (obj->b->length < sizeof(MINIDUMP_HEADER)) {
-		eprintf("Error in r_bin_mdmp_init: length too short, not enough space for MINIDUMP_HEADER\n");
-		return false;
-	}
+  sdb_num_set(obj->kv, "mdmp.hdr.time_date_stamp", obj->hdr->time_date_stamp, 0);
```
Comment (radare) [inline libr/bin/format/mdmp/mdmp.c]: missing spaces before (

File: libr/bin/format/mdmp/mdmp.c [hunk 2]
```diff
@@ -1,371 +1,296 @@
...
+    eprintf("Error: RVA for directory resides in the header!\n");
+    return false;
+  }
 
-	if (obj->hdr->StreamDirectoryRva < sizeof(obj->hdr)) {
-		eprintf("Error in r_bin_mdmp_init: invalid StreamDirectoryRva, size %d\n", obj->hdr->StreamDirectoryRva);
-		return false;
-	}
+  if (obj->hdr->check_sum != 0) {
+    eprintf("TODO: Checksum present but needs validating!\n");
```
Comment (radare) [inline libr/bin/format/mdmp/mdmp.c]: use tabs to indent, not spaces

File: libr/bin/format/mdmp/mdmp.c [hunk 3]
```diff
@@ -1,371 +1,296 @@
...
+  {
+    eprintf("Error: RVA for directory resides in the header!\n");
+    return false;
+  }
 
-	if (obj->hdr->StreamDirectoryRva < sizeof(obj->hdr)) {
-		eprintf("Error in r_bin_mdmp_init: invalid StreamDirectoryRva, size %d\n", obj->hdr->StreamDirectoryRva);
-		return false;
-	}
+  if (obj->hdr->check_sum != 0) {
```
Comment (radare) [inline libr/bin/format/mdmp/mdmp.c]: != is implicit

File: libr/bin/format/mdmp/mdmp.c [hunk 4]
```diff
@@ -1,371 +1,296 @@
...
+  /* We could confirm data sizes but a malcious MDMP will always get around
+   * this! But we can ensure that the data is not outside of the file */
+  if (entry->location.rva + entry->location.data_size > obj->b->length)
+  {
+    eprintf("ERROR: Size Mismatch - Stream data is larger than file size!\n");
+    return false;
+  }
+
+  switch (entry->stream_type) {
+    case THREAD_LIST_STREAM:
```
Comment (radare) [inline libr/bin/format/mdmp/mdmp.c]: cases must be indented at the same level as switch, see r2 indentation style guidelines

File: libr/bin/format/mdmp/mdmp.c [hunk 5]
```diff
@@ -1,371 +1,296 @@
...
+      break;
+    case SYSTEM_INFO_STREAM:
+      obj->streams.system_info = (struct minidump_system_info *)(obj->b->buf + entry->location.rva);
+
+      break;
+    case THREAD_EX_LIST_STREAM:
+      thread_ex_list = (struct minidump_thread_ex_list *)(obj->b->buf + entry->location.rva);
+      for (i = 0; i < thread_ex_list->number_of_threads; i++) {
+        ex_threads = (struct minidump_thread_ex *)(&(thread_ex_list->threads));
+        r_list_append(obj->streams.ex_threads, &(ex_threads[i]));
```
Comment (radare) [inline libr/bin/format/mdmp/mdmp.c]: more spaces before (

File: libr/bin/p/bin_mdmp.c [hunk 6]
```diff
@@ -1,171 +1,292 @@
...
-			obj->system_info->MinorVersion,
-			obj->system_info->BuildNumber);
-		break;
-	default:
-		ret->os = strdup ("unknown");
-	}
-	return ret;
+static int check_bytes(const ut8 *buf, ut64 length) {
+  return buf && (length > sizeof(struct minidump_header))
+    && (memcmp(buf, MDMP_MAGIC, 6) == 0);
```
Comment (radare) [inline libr/bin/p/bin_mdmp.c]: use !memcmp instead of == 0

**Comment by AlexKornitzer** [comment]
That is an error on my account (and my vim setup) :P I will read over the guidelines and rectify ASAP!!!

**Comment by radare** [comment]
cool thanks!

**Comment by AlexKornitzer** [comment]
What are the conventions for breaking generally long lines (over 80 chars)?

1. `static bool r_bin_mdmp_init_directory_entry(struct r_bin_mdmp_obj *obj, struct minidump_directory *entry) {`

2. `handle_operations = (struct avrf_handle_operation *)((ut8 *)handle_operation_list + sizeof(struct minidump_handle_operation_list));`

3. `eprintf("WARNING: <MORE_THAN_80_CHARS> %i\n", entry->stream_type);`

**Comment by AlexKornitzer** [comment]
Guidelines applied!

**Comment by radare** [comment]
Looks good. i'll merge that, but it will be good to have some tests because im not really an MDMP user so i cant say how much better or worst those changes are :D

File: libr/bin/format/mdmp/mdmp.c [hunk 7]
```diff
@@ -1,371 +1,281 @@
-/* radare - LGPL - Copyright 2015 - Dāvis */
+/*
```
Comment (radare) [inline libr/bin/format/mdmp/mdmp.c]: use one liner license statement not this windows-style-ultra-verbose-one

File: libr/bin/format/mdmp/mdmp.c [hunk 8]
```diff
@@ -1,371 +1,281 @@
...
-		sdb_free (obj->kv);
-		obj->kv = NULL;
-	}
-	if (obj->b) {
-		r_buf_free (obj->b);
-		obj->b = NULL;
-	}
-	free (obj);
-}
+/*******************************************************************************
```
Comment (radare) [inline libr/bin/format/mdmp/mdmp.c]: same goes here

File: libr/bin/p/bin_mdmp.c [hunk 9]
```diff
@@ -1,171 +1,288 @@
+/*
```
Comment (radare) [inline libr/bin/p/bin_mdmp.c]: use oneliner copyright comment

**Comment by radare** [comment]
there are still more spaces to be converted into tabs. and change the multi-line license headers. ill merge after that

**Comment by AlexKornitzer** [comment]
Done. On the testing note, I can create some tests using some valid and invalid MDMP files.

File: libr/bin/format/mdmp/mdmp.c [hunk 10]
```diff
@@ -1,371 +1,265 @@
...
+		r_buf_free (obj->b);
+		obj->kv = NULL;
+	}
+	R_FREE (obj);
+
+	return;
+}
+
+struct r_bin_mdmp_obj *r_bin_mdmp_new_buf(struct r_buf_t *buf) {
+	bool streams;
```
Comment (radare) [inline libr/bin/format/mdmp/mdmp.c]: initialize streams in the definition line

File: libr/bin/format/mdmp/mdmp.c [hunk 11]
```diff
@@ -1,371 +1,265 @@
...
+	struct r_bin_mdmp_obj *obj;
+
+	obj = R_NEW0 (struct r_bin_mdmp_obj);
+	obj->kv = sdb_new0 ();
+	obj->b = r_buf_new ();
+	obj->size = (ut32)buf->length;
+
+	streams = true;
+	if (!(obj->streams.ex_threads = r_list_new ())) streams = false;
```
Comment (radare) [inline libr/bin/format/mdmp/mdmp.c]: this seems like a goto emulation what about doing like this?

what about sdoing it like:
```
bool fail = false;
fail |= !(obj->streamds.memories = r_list_new());
if (fail) {...}
```

File: libr/bin/p/bin_mdmp.c [hunk 12]
```diff
@@ -1,171 +1,273 @@
...
+        ret->machine = strdup ("AMD64");
+        ret->arch = strdup ("x86");
+        ret->bits = 64;
+        break;
+      default:
+        strncpy (ret->machine, "Unknown", R_BIN_SIZEOF_STRINGS);
+    }
+
+    switch (obj->streams.system_info->product_type) {
+      case VER_NT_WORKSTATION:
```
Comment (radare) [inline libr/bin/p/bin_mdmp.c]: dont indent with spaces :P

**Comment by radare** [comment]
ther eare still more indentation issues, try with s/^\ \ /\t/g or so

**Comment by AlexKornitzer** [comment]
Third times the charm!? :)

**Comment by AlexKornitzer** [comment]
Honestly unsure as to how I missed that, egg is on my face clearly!

**Pull request merged**